File: Dynamic Programming Applications In Machine Learning and Genomics/Constructing a Profile HMM/profile_hmm.cpp
#include <iostream>
#include <string>
#include <vector>


using std::string;
using std::vector;
using std::cin;
using std::cout;
using std::endl;


/*
 * Creates a profile HMM from a multiple alignment.
 *
 * \param theta         a double threshold for the seed alignment
 * \param sigma         a vector of chars containing the HMM alphabet
 * \param alignment     a vector of strings containing the multiple alignment
 *                          for which the profile HMM will be constructed.
 *
 * \return a string with the emission and transition matrices formatted 
 * according to the problem output specification.
 */
string profile_hmm(double theta, vector<char> sigma, vector<string> alignment) {
    // TODO: your code here
    return "";
}


int main(int argc, char* argv[]) {
    // read in theta threshold
    double theta;
    cin >> theta;
    cin.ignore(1, '\n');    // ignore newline after theta
    cin.ignore(100, '\n');  // delimiter

    // read in HMM alphabet
    vector<char> sigma;
    char letter;
    cin >> letter;
    while (letter != '-') {
        sigma.push_back(letter);
        cin >> letter;
    }
    cin.ignore(100, '\n');  // delimiter

    // read in multiple alignment
    vector<string> alignment;
    string alignment_line;
    while (cin >> alignment_line) {
        alignment.push_back(alignment_line);
    }

    // compute the profile HMM and print it
    cout << profile_hmm(theta, sigma, alignment) << endl;

    return 0;
}

File: Dynamic Programming Applications In Machine Learning and Genomics/Constructing a Profile HMM/ProfileHmm.java
import java.util.Scanner;
import java.util.ArrayList;


public class ProfileHmm {

    /*
     * Creates a profile HMM from a multiple alignment.
     *
     * @param theta         a double threshold for the seed alignment
     * @param sigma         an array of chars containing the HMM alphabet 
     * @param alignment     an array of Strings containing the multiple alignment 
     *                      for which the profile HMM will be constructed.  
     *
     * @return a String with the emission and transition matrices formatted 
     * according to the problem output specification.
     */
    public static String profileHmm(double theta, char[] sigma, String[] alignment) {
        // TODO: your code here
        return "";
    }

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);

        // read in theta threshold
        double theta = stdin.nextDouble();
        stdin.nextLine();   // newline at the end of first
        stdin.nextLine();   // delimiter

        // read in HMM alphabet
        char[] sigma = String.join("", stdin.nextLine().split(" ")).toCharArray();
        stdin.nextLine();   // delimiter

        // read in multiple alignment
        ArrayList<String> tempAlignment = new ArrayList<>(); 
        while (stdin.hasNextLine()) {
            tempAlignment.add(stdin.nextLine());
        }
        String[] alignment = new String[tempAlignment.size()];
        alignment = tempAlignment.toArray(alignment);

        // compute profile HMM
        System.out.println(profileHmm(theta, sigma, alignment));
    }
}
